File: src/libawkward/type/UnionType.cpp
// BSD 3-Clause License; see https://github.com/scikit-hep/awkward-1.0/blob/master/LICENSE

#define FILENAME(line) FILENAME_FOR_EXCEPTIONS("src/libawkward/type/UnionType.cpp", line)

#include <string>
#include <sstream>

#include "awkward/type/UnknownType.h"
#include "awkward/type/OptionType.h"
#include "awkward/array/UnionArray.h"

#include "awkward/type/UnionType.h"

namespace awkward {
  UnionType::UnionType(const util::Parameters& parameters,
                       const std::string& typestr,
                       const std::vector<TypePtr>& types)
      : Type(parameters, typestr)
      , types_(types) { }

  std::string
  UnionType::tostring_part(const std::string& indent,
                           const std::string& pre,
                           const std::string& post) const {
    std::string typestr;
    if (get_typestr(typestr)) {
      return wrap_categorical(typestr);
    }

    std::stringstream out;
    out << indent << pre << "union[";
    for (int64_t i = 0;  i < numtypes();  i++) {
      if (i != 0) {
        out << ", ";
      }
      out << type(i).get()->tostring_part(indent, "", "");
    }
    if (!parameters_empty()) {
      out << ", " << string_parameters();
    }
    out << "]" << post;
    return wrap_categorical(out.str());
  }

  const TypePtr
  UnionType::shallow_copy() const {
    return std::make_shared<UnionType>(parameters_, typestr_, types_);
  }

  bool
  UnionType::equal(const TypePtr& other, bool check_parameters) const {
    if (UnionType* t = dynamic_cast<UnionType*>(other.get())) {
      if (check_parameters  &&  !parameters_equal(other.get()->parameters())) {
        return false;
      }
      if (types_.size() != t->types_.size()) {
        return false;
      }
      for (size_t i = 0;  i < types_.size();  i++) {
        if (!types_[i].get()->equal(t->types_[i], check_parameters)) {
          return false;
        }
      }
      return true;
    }
    else {
      return false;
    }
  }

  int64_t
  UnionType::numtypes() const {
    return (int64_t)types_.size();
  }

  int64_t
  UnionType::numfields() const {
    throw std::runtime_error(
      std::string("FIXME: UnionType::numfields") + FILENAME(__LINE__));
  }

  int64_t
  UnionType::fieldindex(const std::string& key) const {
    throw std::runtime_error(
      std::string("FIXME: UnionType::fieldindex(key)") + FILENAME(__LINE__));
  }

  const std::string
  UnionType::key(int64_t fieldindex) const {
    throw std::runtime_error(
      std::string("FIXME: UnionType::key(fieldindex)") + FILENAME(__LINE__));
  }

  bool
  UnionType::haskey(const std::string& key) const {
    throw std::runtime_error(
      std::string("FIXME: UnionType::haskey(key)") + FILENAME(__LINE__));
  }

  const std::vector<std::string>
  UnionType::keys() const {
    throw std::runtime_error(
      std::string("FIXME: UnionType::keys") + FILENAME(__LINE__));
  }

  const std::vector<TypePtr>
  UnionType::types() const {
    return types_;
  }

  const ContentPtr
  UnionType::empty() const {
    ContentPtrVec contents;
    for (auto type : types_) {
      contents.push_back(type.get()->empty());
    }
    Index8 tags(0);
    Index64 index(0);
    return std::make_shared<UnionArray8_64>(Identities::none(),
                                            parameters_,
                                            tags,
                                            index,
                                            contents);
  }

  const TypePtr
  UnionType::type(int64_t index) const {
    return types_[(size_t)index];
  }
}

File: src/awkward/_v2/types/uniontype.py
# BSD 3-Clause License; see https://github.com/scikit-hep/awkward-1.0/blob/main/LICENSE

from collections.abc import Iterable

import awkward as ak
from awkward._v2.types.type import Type
from awkward._v2.forms.form import _parameters_equal


class UnionType(Type):
    def __init__(self, contents, parameters=None, typestr=None):
        if not isinstance(contents, Iterable):
            raise ak._v2._util.error(
                TypeError(
                    "{} 'contents' must be iterable, not {}".format(
                        type(self).__name__, repr(contents)
                    )
                )
            )
        if not isinstance(contents, list):
            contents = list(contents)
        for content in contents:
            if not isinstance(content, Type):
                raise ak._v2._util.error(
                    TypeError(
                        "{} all 'contents' must be Type subclasses, not {}".format(
                            type(self).__name__, repr(content)
                        )
                    )
                )
        if parameters is not None and not isinstance(parameters, dict):
            raise ak._v2._util.error(
                TypeError(
                    "{} 'parameters' must be of type dict or None, not {}".format(
                        type(self).__name__, repr(parameters)
                    )
                )
            )
        if typestr is not None and not ak._util.isstr(typestr):
            raise ak._v2._util.error(
                TypeError(
                    "{} 'typestr' must be of type string or None, not {}".format(
                        type(self).__name__, repr(typestr)
                    )
                )
            )
        self._contents = contents
        self._parameters = parameters
        self._typestr = typestr

    @property
    def contents(self):
        return self._contents

    def _str(self, indent, compact):
        if self._typestr is not None:
            out = [self._typestr]

        else:
            if compact:
                pre, post = "", ""
            else:
                pre, post = "\n" + indent + "    ", "\n" + indent

            children = []
            for i, x in enumerate(self._contents):
                if i + 1 < len(self._contents):
                    if compact:
                        y = x._str(indent, compact) + [", "]
                    else:
                        y = x._str(indent + "    ", compact) + [",\n", indent, "    "]
                else:
                    if compact:
                        y = x._str(indent, compact)
                    else:
                        y = x._str(indent + "    ", compact)
                children.append(y)

            flat_children = [y for x in children for y in x]
            params = self._str_parameters()

            if params is None:
                out = ["union[", pre] + flat_children + [post, "]"]
            else:
                out = ["union[", pre] + flat_children + [", ", post, params, "]"]

        return [self._str_categorical_begin()] + out + [self._str_categorical_end()]

    def __repr__(self):
        args = [repr(self._contents)] + self._repr_args()
        return "{}({})".format(type(self).__name__, ", ".join(args))

    def __eq__(self, other):
        if isinstance(other, UnionType):
            return (
                _parameters_equal(
                    self._parameters, other._parameters, only_array_record=True
                )
                and self._contents == other._contents
            )
        else:
            return False
